### ISSUE TITLE
GRPC extremely slow depending on model Response 

### ISSUE THREAD

**Issue by mmeendez8** [pytorch/serve#2093]
### 🐛 Describe the bug

I have benchmarked GRPC vs HTTP expecting to achieve certain improvements on my application but I was surprised to find that GRPC can be a lot slower compared to HTTP.

Basically GPRC performance is highly dependent on model response. This is probably due to[ .proto declaration](https://github.com/pytorch/serve/blob/master/frontend/server/src/main/resources/proto/inference.proto#L20-L23):

```
message PredictionResponse {
    // Response content for prediction
    bytes prediction = 1;
}
```

The whole response JSON gets encoded as bytes which is very inefficient

### Error logs

Model that returns top 1 label:

```
HTTP Mean time  : 2.559002089500427
GRPC Mean time  : 2.6132653951644897
```

Model that returns top 100 labels:

```
HTTP Mean time  : 2.608074736595154
GRPC Mean time  : 2.8387032985687255
```

Note this results are using localhost where we have a huge bandwith. When I reproduce this in my k8s cluster I found large differences, http is a few seconds faster than gRPC.

### Installation instructions

Just run latest docker gpu image.

### Model Packaing

* Handler: `ImageClassifier `handler but modified topk to 100:

```python
import torch
import torch.nn.functional as F
from torchvision import transforms

from ts.torch_handler.vision_handler import VisionHandler
from ts.utils.util  import map_class_to_label


class ImageClassifier(VisionHandler):
    """
    ImageClassifier handler class. This handler takes an image
    and returns the name of object in that image.
    """

    topk = 100
    # These are the standard Imagenet dimensions
    # and statistics
    image_processing = transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406],
                             std=[0.229, 0.224, 0.225])
    ])

    def set_max_result_classes(self, topk):
        self.topk = topk

    def get_max_result_classes(self):
        return self.topk

    def postprocess(self, data):
        ps = F.softmax(data, dim=1)
        probs, classes = torch.topk(ps, self.topk, dim=1)
        probs = probs.tolist()
        classes = classes.tolist()
        return [map_class_to_label(probs, self.mapping, classes)] # replace this with {} for checking with empty response
```

* Model file:

```bash
curl -o example/model.py https://raw.githubusercontent.com/pytorch/serve/master/ts/torch_handler/image_classifier.py
```

* Weights:

```bash
curl -o example/densenet161-8d451a50.pth https://download.pytorch.org/models/densenet161-8d451a50.pth
```

* Labels dict:

```bash
curl -o example/index_to_name.json https://raw.githubusercontent.com/pytorch/serve/master/examples/image_classifier/index_to_name.json
```

* Model Archiver command:

```bash
torch-model-archiver --model-name densenet161 --version 1.0 --model-file example/model.py --serialized-file example/densenet161-8d451a50.pth --export-path example/model-store --extra-files example/index_to_name.json --handler example/custom_handler.py
```

### config.properties

Didn't use, default configs.

### Versions

```
------------------------------------------------------------------------------------------
Environment headers
------------------------------------------------------------------------------------------
Torchserve branch: 

torchserve==0.5.3
torch-model-archiver==0.5.3

Python version: 3.8 (64-bit runtime)
Python executable: /home/mmendez/.cache/pypoetry/virtualenvs/ai-grpc-benchmark-lmaOpNG--py3.8/bin/python

Versions of relevant python libraries:
captum==0.5.0
numpy==1.24.1
psutil==5.9.4
requests==2.28.2
torch==1.11.0+cu113
torch-model-archiver==0.5.3
torchserve==0.5.3
torchvision==0.12.0+cu113
wheel==0.37.1
torch==1.11.0+cu113
**Warning: torchtext not present ..
torchvision==0.12.0+cu113
**Warning: torchaudio not present ..

Java Version:


OS: Ubuntu 20.04.3 LTS
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: N/A
CMake version: N/A

Is CUDA available: Yes
CUDA runtime version: N/A
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3070 Laptop GPU
Nvidia driver version: 516.94
cuDNN version: None
```

### Repro instructions

Run torchserve with:

```bash
docker run -it --rm --shm-size=1g \
        --ulimit memlock=-1 \
        --ulimit stack=67108864 \
        -p8080:8080 \
        -p8081:8081 \
        -p8082:8082 \
        -p7070:7070 \
        -p7071:7071 \
        --mount type=bind,source=/home/mmendez/work/ai_grpc_benchmark/example/model-store,target=/tmp/models pytorch/torchserve:latest-gpu torchserve --model-store=/tmp/models --start --models all

```

Run the following benchmark script:

```python
import cv2
import numpy as np
from time import time
from httpx import AsyncClient, Timeout

import grpc
from ai_grpc_benchmark.ts_stubs import inference_pb2, inference_pb2_grpc
import asyncio

async def grpc_call(stub, model_name, img_bytes):
    request_data = inference_pb2.PredictionsRequest(
            model_name=model_name,
            input={"data": img_bytes},
        )
    return await stub.Predictions(request_data)


async def benchmark():
    http_url = "http://localhost:8080"
    grpc_url = "localhost:7070"
    model_name = "densenet161"

    img = np.ones((1080, 1980, 3), dtype = np.uint8)
    img_bytes = cv2.imencode(".jpg", img)[1].tobytes()
    n_requests = 10
    concurrent_request = 50

    # Send HTTP requests
    client = AsyncClient(timeout=Timeout(timeout=30.0))
    http_times = []
    for _ in range(n_requests):
        start = time()

        tasks = []
        for _ in range(concurrent_request):
            tasks.append(asyncio.create_task(client.post(url=f"{http_url}/predictions/{model_name}", data=img_bytes)))
        response = await asyncio.gather(*tasks)
        end = time() - start
        http_times.append(end)
        print(f'HTTP call took \t: {end}')

    # Send GRPC requests
    channel = grpc.aio.insecure_channel(grpc_url)

    stub = inference_pb2_grpc.InferenceAPIsServiceStub(channel)

    grpc_times = []
    for _ in range(n_requests):

        start = time()

        tasks = []
        for _ in range(concurrent_request):
            tasks.append(asyncio.create_task(grpc_call(stub, model_name, img_bytes)))
        
        response = await asyncio.gather(*tasks)

        end = time() - start
        grpc_times.append(end)
        print(f'GRPC call took \t: {end}')

    print("\nSummary:\n" + 100*"=")
    print(f'HTTP Mean time \t: {np.mean(np.array(http_times))}')
    print(f'GRPC Mean time \t: {np.mean(np.array(grpc_times))}')


if __name__ == '__main__':

    asyncio.run(benchmark())
```

### Possible Solution

Find a way of adding user customized Response types for gRPC.

**Comment by agunapal** [comment]
Thanks for reporting this @mmeendez8  What do you mean by its  highly dependent on model response? Do you mind elaborating where it is slow and where it works as expected

**Comment by mmeendez8** [comment]
Hi @agunapal, sorry for the last response! I was trying to find time to do a proper analysis, capture all my thoughts in this post:
[https://mmeendez8.github.io/2022/02/24/torchserve-grpc.html](https://mmeendez8.github.io/2022/02/24/torchserve-grpc.html)

What I was trying to say in here is that .proto prediction response is:

```python
message PredictionResponse {
    // Response content for prediction
    bytes prediction = 1;
}
```
And I think that cannot be customized that so you are not getting any of the protobuf advantages for you model responses.

**Comment by agunapal** [comment]
@mmeendez8 Thanks for the amazing post and feedback!  I have shared it with the team.
